clients/src/main/java/org/apache/kafka/common/network/KafkaChannel.java

```diff
@@ -28,6 +28,52 @@
 import java.util.Objects;
 
 public class KafkaChannel {
+    /**
+     * Mute States for KafkaChannel:
+     * <ul>
+     *   <li> NOT_MUTED: Channel is not muted. This is the default state. </li>
+     *   <li> MUTED: Channel is muted. Channel must be in this state to be unmuted. </li>
+     *   <li> MUTED_AND_RESPONSE_PENDING: (SocketServer only) Channel is muted and currently waiting to receive a
+     *                                   response from the API layer. </li>
+     *   <li> MUTED_AND_THROTTLED: (SocketServer only) Channel is muted and throttling is in progress due to quota
+     *                             violation. </li>
+     *   <li> MUTED_AND_THROTTLED_AND_RESPONSE_PENDING: (SocketServer only) Channel is muted, throttling is in progress,
+     *                                                 and waiting for a response from the API layer. </li>
+     * </ul>
+     */
+    public enum ChannelMuteState {
+        NOT_MUTED,
+        MUTED,
+        MUTED_AND_RESPONSE_PENDING,
+        MUTED_AND_THROTTLED,
+        MUTED_AND_THROTTLED_AND_RESPONSE_PENDING
+    };
+
+    /** Socket server events that will change the mute state. Valid transitions are:
+     * <ul>
+     *   <li> REQUEST_SENT: A request has been sent to the API layer. </li>
+     *   <li> RESPONSE_RECEIVED: A response has been received from the API layer. </li>
+     *   <li> THROTTLE_STARTED: Throttling started due to quota violation. </li>
+     *   <li> THROTTLE_ENDED: Throttling ended. </li>
+     * </ul>
+     *
+     * Valid transitions on each event are:
+     * <ul>
+     *   <li> REQUEST_SENT:      MUTED => MUTED_AND_RESPONSE_PENDING </li>
+     *   <li> RESPONSE_RECEIVED: MUTED_AND_RESPONSE_PENDING => MUTED, MUTED_AND_THROTTLED_AND_RESPONSE_PENDING =>
+     *                           MUTED_AND_THROTTLED </li>
+     *   <li> THROTTLE_STARTED:  MUTED_AND_RESPONSE_PENDING => MUTED_AND_THROTTLED_AND_RESPONSE_PENDING </li>
+     *   <li> THROTTLE_ENDED:    MUTED_AND_THROTTLED => MUTED, MUTED_AND_THROTTLED_AND_RESPONSE_PENDING =>
+     *                           MUTED_AND_RESPONSE_PENDING </li>
+     * </ul>
+     */
+    public enum ChannelMuteEvent {
+        REQUEST_SENT,
+        RESPONSE_RECEIVED,
```

Comment:
REQUEST/RESPONSE in this enum and the previous one refer to request/response between SocketServer and the API layer, not between the client and SocketServer. One reason why I chose this way was because we don't send out a response to the client when acks = 0.

But you are right that we try unmuting after sending out to the response to the client, so I updated as suggested. I still use RESPONSE_SENT for acks=0 case, but I think it should not be confusing (mentioned that in the comments).